source/common/grpc/common.cc

```diff
@@ -146,20 +146,86 @@ Buffer::InstancePtr Common::serializeBody(const Protobuf::Message& message) {
+  if (header_grpc_timeout_entry) {
+    uint64_t grpc_timeout;
+    const char* unit =
+        StringUtil::strtoul(header_grpc_timeout_entry->value().c_str(), grpc_timeout);
+    if (unit != nullptr && *unit != '\0') {
+      switch (*unit) {
+      case 'H':
+        timeout = std::chrono::hours(grpc_timeout);
+        break;
+      case 'M':
+        timeout = std::chrono::minutes(grpc_timeout);
+        break;
+      case 'S':
+        timeout = std::chrono::seconds(grpc_timeout);
+        break;
+      case 'm':
+        timeout = std::chrono::milliseconds(grpc_timeout);
+        break;
+      case 'u':
+        timeout = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::microseconds(grpc_timeout));
+        if (timeout < std::chrono::microseconds(grpc_timeout)) {
+          timeout++;
+        }
+        break;
+      case 'n':
+        timeout = std::chrono::duration_cast<std::chrono::milliseconds>(
+            std::chrono::nanoseconds(grpc_timeout));
+        if (timeout < std::chrono::nanoseconds(grpc_timeout)) {
+          timeout++;
+        }
+        break;
+      }
+    }
+  }
+  return timeout;
+}
+
+static void toGrpcTimeout(const std::chrono::milliseconds& timeout, Http::HeaderString& value) {
+  uint64_t time = timeout.count();
+  static const char units[] = "mSMH";
+  const char* unit = units; // start with milliseconds
+  static constexpr size_t MAX_GRPC_TIMEOUT_VALUE = 99999999;
+  if (time > MAX_GRPC_TIMEOUT_VALUE) {
+    time /= 1000; // Convert from milliseconds to seconds
+    unit++;
+  }
+  while (time > MAX_GRPC_TIMEOUT_VALUE) {
+    if (*unit == 'H') {
+      time = MAX_GRPC_TIMEOUT_VALUE; // No bigger unit available, clip to max 8 digit hours.
+    } else {
+      time /= 60; // Convert from seconds to minutes to hours
+      unit++;
+    }
+  }
+  value.setInteger(time);
+  value.append(unit, 1);
+}
+
 Http::MessagePtr Common::prepareHeaders(const std::string& upstream_cluster,
                                         const std::string& service_full_name,
-                                        const std::string& method_name) {
+                                        const std::string& method_name,
+                                        const absl::optional<std::chrono::milliseconds>& timeout) {
   Http::MessagePtr message(new Http::RequestMessageImpl());
   message->headers().insertMethod().value().setReference(Http::Headers::get().MethodValues.Post);
   message->headers().insertPath().value().append("/", 1);
   message->headers().insertPath().value().append(service_full_name.c_str(),
                                                  service_full_name.size());
   message->headers().insertPath().value().append("/", 1);
   message->headers().insertPath().value().append(method_name.c_str(), method_name.size());
+  message->headers().insertTE().value().setReference(Http::Headers::get().TEValues.Trailers);
```

Comment:
Why this move?

Reply:
It is a remnant from when I did not yet fully grasp how ordering of headers works, will revert.

gRPC spec requires the TE header to appear before the Authority header, even though "te" does not start with ':', with a comment "*Used to detect incompatible proxies*". In my testing I found out, however, that nghttp2 does not allow that, meaning that if the gRPC endpoints care to check for it, Envoy will appear to them as an "incompatible proxy". I have no idea if this has real world implications, though.

Reply:
ON a second thought, maybe we should not simply revert this, but move TE as early as possible (first non-pseudo header) instead?